cnd/src/btsieve/ethereum/web3_connector.rs

```diff
@@ -9,32 +9,60 @@ use crate::{
         BlockId, BlockNumber,
     },
 };
+use async_std::sync::Mutex;
+use async_trait::async_trait;
 use futures::Future;
+use futures_core::{compat::Future01CompatExt, TryFutureExt};
+use lru_cache::LruCache;
 use reqwest::Url;
 use std::sync::Arc;
 
+type Block = Option<crate::ethereum::Block<crate::ethereum::Transaction>>;
+type Hash = crate::ethereum::H256;
+
 #[derive(Clone, Debug)]
 pub struct Web3Connector {
     web3: Arc<Web3<Http>>,
     task_executor: tokio::runtime::TaskExecutor,
+    block_cache: Web3BlockCache,
 }
 
 impl Web3Connector {
     pub fn new(
         node_url: Url,
         task_executor: tokio::runtime::TaskExecutor,
+        cache_capacity: usize,
     ) -> Result<(Self, EventLoopHandle), web3::Error> {
         let (event_loop_handle, http_transport) = Http::new(node_url.as_str())?;
         Ok((
             Self {
                 web3: Arc::new(Web3::new(http_transport)),
                 task_executor,
+                block_cache: Web3BlockCache::new(cache_capacity),
             },
             event_loop_handle,
         ))
     }
 }
 
+block_cache!(Web3BlockCache);
+
+#[async_trait]
+impl BlockCache for Web3BlockCache {
+    type Block = Block;
+    type BlockHash = Hash;
+
+    async fn get(&self, block_hash: &Hash) -> anyhow::Result<Option<Block>> {
+        let mut cache = self.map.lock().await;
+        Ok(cache.get_mut(block_hash).cloned())
+    }
+
+    async fn insert(&mut self, block_hash: Hash, block: Block) -> anyhow::Result<Option<Block>> {
```

Comment:
The `Map` is wrapped in a `Mutex`, you don't need `&mut self`. I am also not sure whether interior mutability is good here. We could also just have an `Arc<Mutex<BlockCache>>`. That would make it more explicit that the cache is shared.